### ISSUE TITLE
Seeking feedback on new API for adaptive acquisition

### ISSUE THREAD

**Issue by henrypinkard** [micro-manager/pycro-manager#654]
Following up on https://github.com/micro-manager/pycro-manager/issues/567

Making an asynchronous equivalent of what hooks, image processors, and image saved callbacks do right now would make it much easier to express complex experiments in a readable way in pycromanager. It would provide the benefits of the parallelized backend (i.e. hardware and images handled on separate threads), but would allow user code to be readable, sequential commands.

I think I know approximately how to do this on the backend, but it would be great to get some feedback on what the API should look like. This is powered by adding to the acquisition engine a new feature I'm calling "notifications": asynchronous updates about what is going on on its different threads (e.g. finished the `pre_hardware_hook` for `{'time': 3, 'z': 4}`, acquired image `'{'time': 5, 'z': 6}'`, etc.). 

One way of doing this from the user perspective, when you call `acq.acquire(events)`, it would now return an `AcquisitionFuture` object that allows you to wait on these various things happening. Example:

```python
with Acquisition(...) as acq:
   events = multi_d_acquisition_events(num_time_points=5)
   af = acq.acquire(events)
   # now pause until the 3rd time point is about  to be acquired
   # unlike a hook, this call will not block the acq engine from proceeding after it reaches this point
   # it just blocks the main thread from proceeding until this has happened
   af.await_execution(milestone='pre_hardware', time=3) 

  # or a different way of doing this would be to only keep the AcquisitionFuture internally
  # and just call 
  acq.await_execution(milestone='pre_hardware', time=3) 

```

Could also have this for images (and this might even be more useful than hooks for data-adaptive acquisitions):

```python
with Acquisition(...) as acq:
   events = multi_d_acquisition_events(num_time_points=5)
   af = acq.acquire(events)
   image = af.await_image_saved(return_image=True, time=3)
   # process the image here and decide what events to send next
```

I think it probably would not make sense to have a `await_image_acquired(return_image=False)`, because this would mean all images would need to be hung onto in RAM in case you called this function later.

Behind the scenes, when you call `acq.acquire`, the `AcquisitionFuture` is dynamically generated with knowledge of all possible "milestones" it could be awaiting (i.e. all points in the acquisition cycle for all `axes` combos present in the events). As long as the `AcquisitionFuture` is not garbage collected, it is monitoring the notifications coming from the acquisition engine, so that it can knows that a notification has been received even if you ask it to `await_execution` after the notification already happened. `AcquisitionFuture`s going out of scope and getting garbage collected is how you keep notifications from creating too many objects and clogging up memory
 
How should this API look?

 - acq.awaitExecution() or explicitly get an `AcquisitionFuture` object? One advantage I see of the latter would be that it might be able to be dynamically generated in some smart way to have attributes corresponding to all the things it can wait on, like `af.time`, `af.z` etc. Is this useful or are there better things that could be done?
 - Is there a better keyword than `milestone` to express "point in the hardware control and image acquisition cycle"?
 - Other ideas or improvements?

**Comment by henrypinkard** [comment]
example:

```python
from pycromanager import Acquisition, multi_d_acquisition_events
from pycromanager.notifications import AcqNotification
import time
import numpy as np

events = multi_d_acquisition_events(num_time_points=10, time_interval_s=2)

start = time.time()
with Acquisition(directory='/Users/henrypinkard/tmp', name='acq', show_display=False) as acq:
    future = acq.acquire(events)
    
    future.await_execution({'time': 5}, AcqNotification.Hardware.POST_HARDWARE)
    print('time point 5 post hardware')
    
    image = future.await_image_saved({'time': 5}, return_image=True)
    print('time point 5 image saved ', time.time() - start,  '\t\tmean image value: ', np.mean(image))
    
    future.await_execution({'time': 8}, AcqNotification.Hardware.POST_HARDWARE)
    print(time.time() - start, ' time point 8 post hardware')

```

@ieivanov Interested in trying this out and seeing what you think? I think that has the potential to greatly simplify the way you currently do this in #567. You can try this out with latestest nightly build + pulling the branch I have in this PR: https://github.com/micro-manager/pycro-manager/actions/runs/5971850524/job/16201520508?pr=677

**Comment by henrypinkard** [comment]
@wl-stepp maybe this is of interest to you as well?

**Comment by wl-stepp** [comment]
Sounds interesting, thanks. Will have a closer look and think about it.

**Comment by ieivanov** [comment]
Hey @henrypinkard, thanks for your work on this. I'm a bit tied up with work on a manuscript for the next two months. I expect I won't be doing much software development until then, but I'll be eager to try it after!

**Comment by henrypinkard** [comment]
@ieivanov  No worries, good luck with the paper!

**Comment by wl-stepp** [comment]
>     * acq.awaitExecution() or explicitly get an `AcquisitionFuture` object? One advantage I see of the latter would be that it might be able to be dynamically generated in some smart way to have attributes corresponding to all the things it can wait on, like `af.time`, `af.z` etc. Is this useful or are there better things that could be done?

Feels natural to explicitly get the `AcquisitionFuture`. How does it work if acquisition events are added later during a hook, or after waiting for a milestone? Can I get a new updated future that allows me to wait for the newly added event?
I like that it doesn't block, but what happens if the acquisition ends while I'm doing things on something I waited for? Is there a way of keeping the acquisition active even if there are no more events?

>     * Is there a better keyword than `milestone` to express "point in the hardware control and image acquisition cycle"?

Milestone sounds intuitive. Acquisition milestone maybe if it's needed to be more explicit?

**Comment by jacopoabramo** [comment]
@henrypinkard just chiming in since this is crossing paths with my work on [this PR](https://github.com/ImSwitch/ImSwitch/pull/177). The notification system would be ideal for me to monitor the acquisition progression and show it on the UI interactively. I tried testing the snippet above with `Acquisition.ACQ_EVENTS_FINISHED` with the following snippet:

```py
from pycromanager import Acquisition, multi_d_acquisition_events, AcqNotification, start_headless
from pymmcore_plus import find_micromanager
import time, os
import numpy as np

events = multi_d_acquisition_events(num_time_points=10, time_interval_s=0)

mm_path = find_micromanager()
config_file = os.path.join(mm_path, 'MMConfig_demo.cfg')

print(config_file)

start_headless(mm_app_path=mm_path, config_file=config_file)

start = time.time()
with Acquisition(directory='./test_rec', name='acq_notification', show_display=False) as acq:
    start = time.time()
    future = acq.acquire(events)

    future.await_execution({'time': 5}, AcqNotification.Acquisition.ACQ_EVENTS_FINISHED)
    print('Acquisition finished')
```

```
Traceback (most recent call last):
  File "C:\git\sandbox\pycromanager_notification_test.py", line 23, in <module>
    future.await_execution({'time': 5}, AcqNotification.Acquisition.ACQ_EVENTS_FINISHED)
  File "C:\Users\jacop\AppData\Local\Programs\Python\Python310\lib\site-packages\pycromanager\acq_future.py", line 58, in await_execution
    notification = AcqNotification(None, axes, phase)
  File "C:\Users\jacop\AppData\Local\Programs\Python\Python310\lib\site-packages\pycromanager\acquisition\acq_eng_py\main\acq_notification.py", line 58, in __init__ 
    raise ValueError("Unknown phase")
ValueError: Unknown phase
```

I imagine this is due to the fact that the implementation is still missing. One thing I would like to have though - and this probably extends to all hook functions - would be the possibility to have a list of callables rather than a single one; in my case, it's to make sure that the basic acquisition widget functionalities can be expanded in case a widget that uses PycroManager can implement another hook function of the same type.

**Comment by wl-stepp** [comment]
> So it might be simplest to say that if you're adding events from a hook, you cant get access to these objects. Do you have a specific use case you're thinking of here?

Well, our event-driven acquisitions would add events and ideally we don't have to specify how many beforehand. Then if something else is done with the futures, say every 20 frames, it might not be possible to await it if the event was not there in the beginning. Do I get that correctly?

`phase` sounds to me more like an extended thing, not a point in the acquisition. But not a native speaker myself...

**Comment by henrypinkard** [comment]
> One thing I would like to have though - and this probably extends to all hook functions - would be the possibility to have a list of callables rather than a single one; in my case, it's to make sure that the basic acquisition widget functionalities can be expanded in case a widget that uses PycroManager can implement another hook function of the same type.

@jacopoabramo no problem to add a list of callables as an option as well.

And yes there are definitely still some bugs to work out here...This isn't fully tested or finalized yet

**Comment by henrypinkard** [comment]
> `phase` sounds to me more like an extended thing, not a point in the acquisition. But not a native speaker myself...

@wl-stepp This is great feedback! I think I will change it to milestone


> Well, our event-driven acquisitions would add events and ideally we don't have to specify how many beforehand. Then if something else is done with the futures, say every 20 frames, it might not be possible to await it if the event was not there in the beginning. Do I get that correctly?

Maybe the solution here is to remove the version of the hook function where you get access to the event queue, and instead you'd just accomplish the same thing by calling `acq.acquire` within the hook.

A big motivation for this API is to make things like EDA easier, such that you can write a single script specifying what you're trying to do that makes the feedback loops more clear, while still getting performance benefits of parallelization, simple scripting, etc. For example:


```python
with Acquisition...as acq:
   while True:
     af = acq.acquire(multi_d_acquisition_events(num_time_points=1))
     image = af.await_image_saved(return_image=True, time=0)

     # analyze the image
     if interesting_phenotype:
       # run a fast timelapse
        events = multi_d_acquisition_events(num_time_points=10, time_interval_s=0)
        af = acq.acquire(events)
     else:
        time.sleep(10) # slow timelapse

```

So the hope would be that you can accomplish everything you need to without having to add events in the hook. What do you think?

**Comment by henrypinkard** [comment]
Thanks for the feedback on notifications @jacopoabramo !



> I'm not sure if this is intentional but, if possible, I would like to have a `POST_EXPOSURE` notification with it's relative integer ID, for each captured image.

Unfortunately this is not possible with the current APIs. In this case of a timelapse with no delay in between, the acquisition engine runs a sequence acquisition, which allows the camera to capture frames faster. However, there is simply a start sequence API call and a stop sequence one, without any intermediate calls or updates from the hardware coming from individual frames having the exposure finish.

So, confusingly, the notifications are currently different depending on whether the acquisition engine is executing a series of snaps or a sequence. The new camera API (https://github.com/micro-manager/mmCoreAndDevices/issues/243) can address this and provide these notifications in all cases, but there will still likely be many cameras operating on the old API for some time. I think I'm going to change the notification types to `pre_snap` \ `post_snap` and `pre_sequence` \ `post_sequence` to avoid this confusion


> This is more coherent with what I expect from the previous use-case; what I would like to have is an integer ID of the current captured position (i.e. position_id), in order to better monitor the execution of the acquisition.

I'm confused what you mean for this one. Could you not just accomplish this by not passing labels for the positions?

> Not sure if it is intentional or a bug, but apparently the z ID never changes.

That's because the notifications are printing the `axes` rather than the full event, since sending the full event in the notification might come with a performance penalty, and presumably the user already has the event anyway.

That XYZ positions feature is designed to support the case where you move to different positions in 3D and then run z-stacks. Here, there are no Z-stacks, so every position has a `z` index of 0. I think it would be reasonable to change the `multi_d_acq_events` function to not include `z` in the axes in the case where z-stacks are not in use if you have any interest in PRing that

```python
 {'axes': {'time': 2, 'position': 'P1', 'z': 0}, 'x': 100, 'y': 0, 'z': 50},
 {'axes': {'time': 2, 'position': 'P2', 'z': 0}, 'x': 0, 'y': 100, 'z': 50},
 {'axes': {'time': 2, 'position': 'P3', 'z': 0}, 'x': 100, 'y': 100, 'z': 100}]
```

**Comment by jacopoabramo** [comment]
@henrypinkard 

> Unfortunately this is not possible with the current APIs.

Got it; I think I managed somehow to solve this issue in my implementation but I honestly can't recall how I dealt with it. 

> I'm confused what you mean for this one. Could you not just accomplish this by not passing labels for the positions?

The reason why I asked this was because in my ImSwitch PR I wanted to monitor an acquisition over XYZ using indexes that could be used by a Qt progress bar; in the end I opted for generating the indexes directly into ImSwitch and parse the incoming position label to retrieve the correct index (the assumption being that pycro-manager would perform each step sequentially), so it's not important anymore.

> I think it would be reasonable to change the multi_d_acq_events function to not include z in the axes in the case where z-stacks are not in use if you have any interest in PRing that

Can't promise I'll have the time to prepare a PR but I'll have a look at the code for sure

**Comment by henrypinkard** [comment]
All sounds good. No pressure on making a PR -- only if you want to
